### PR TITLE
Optimize tokenization loop for very large corpora

### PR DESCRIPTION
**Pull request by koheiw**
This is for restructuring of tokens internals to increase the efficiency with very large corpora.

- Concentrate slow functions within a `lapply()` loop (which can be replaced by `future_lapply()`) (#1965)
- Change how tokenization progress is displayed
- Move character normalization (e.g. hyphens) to tokenizer
- Fix handling of the `verbose` option in `tokens.tokens()` (#1926)

The internal design is tested in `tokens_parallel()` in [a separate branch](https://github.com/quanteda/quanteda/blob/dev-tokens_parallel/R/tokens_parallel.R) and worked with a corpus of 30 million texts (20GB) (takes only 30 min in serial process).

Remaining questions are

- How to print the progress when `verbose = TRUE`?
- When should be switch to `future_lapply()`?

### CODE REVIEW

**Comment by koheiw** [comment]
Speed is more or less the same for the medium size Guardian corpus
```r
 > length(txt) 
[1] 204061

# this branch
> microbenchmark::microbenchmark(
+ tokens(txt, verbose = FALSE), times = 5
+ )
Unit: seconds
                         expr      min       lq     mean   median       uq      max neval
 tokens(txt, verbose = FALSE) 313.8735 315.4183 317.3557 317.2339 319.5595 320.6934     5

Restarting R session...


# master
> microbenchmark::microbenchmark(
+ tokens(txt, verbose = FALSE), times = 5
+ )
Unit: seconds
                         expr      min       lq     mean   median       uq      max neval
 tokens(txt, verbose = FALSE) 302.9805 304.0984 310.1614 305.5729 315.5114 322.6437     5
 

```

File: R/tokens.R [hunk 1]
```diff
@@ -264,42 +264,44 @@ tokens.corpus <- function(x,
...
-        if (verbose) catm(" ...tokenizing", i, "of", length(x), "blocks\n")
-        temp <- tokenizer_fn(x[[i]], split_hyphens = split_hyphens, verbose = verbose)
-        if (verbose) catm(" ...serializing tokens, ")
-        if (i == 1) {
-            x[[i]] <- serialize_tokens(temp)
-        } else {
-            x[[i]] <- serialize_tokens(temp, attr(x[[i - 1]], "types"))
+    x <- texts(x)
+    x <- split(x, factor(ceiling(seq_along(x) / 10000)))
+    x <- lapply(x, function(y) {
```
Comment (koheiw) [inline R/tokens.R:273]: This can be `future_lapply`.

File: R/tokenizers.R [hunk 2]
```diff
@@ -38,7 +38,7 @@ NULL
 #' @export
 tokenize_word <- function(x, split_hyphens = FALSE, verbose = quanteda_options("verbose")) {
     
-    if (verbose) catm(" ...segmenting tokens\n")
+    if (verbose) catm(" ...segmenting texts\n")
```
Comment (kbenoit) [inline R/tokenizers.R:41]: perhaps: `" ...segmenting words\n"` ?

File: R/tokens.R [hunk 3]
```diff
@@ -264,42 +264,44 @@ tokens.corpus <- function(x,
...
-    for (i in seq_along(x)) {
-        if (verbose) catm(" ...tokenizing", i, "of", length(x), "blocks\n")
-        temp <- tokenizer_fn(x[[i]], split_hyphens = split_hyphens, verbose = verbose)
-        if (verbose) catm(" ...serializing tokens, ")
-        if (i == 1) {
-            x[[i]] <- serialize_tokens(temp)
-        } else {
-            x[[i]] <- serialize_tokens(temp, attr(x[[i - 1]], "types"))
+    x <- texts(x)
+    x <- split(x, factor(ceiling(seq_along(x) / 10000)))
```
Comment (kbenoit) [inline R/tokens.R:272]: Doesn't have to be done here, but at some point the fixed 10000 might best to move to an option.

**Comment by kbenoit** [comment]
Oh and on the other qn: Let's merge this first and work on the `future_lapply()` next. That way we isolate the parallelism issues and can test that throughly on different platforms.

File: R/tokenizers.R (hunk 2 above)
Comment (koheiw) [reply R/tokenizers.R:41]: Change to `segmenting into words`.

File: R/tokens.R (hunk 3 above)
Comment (koheiw) [reply R/tokens.R:272]: Added a quanteda option `block_size`

**Comment by koheiw** [comment]
I hide  process ID from the verbose message for now because it is useful only in parallel tokenization. "print the verbose stuff on just the first pass" is possible only in serial tokenization. We can work on these in parallelization branch.

File: R/tokens.R (hunk 3 above)
Comment (kbenoit) [reply R/tokens.R]: 👍  but how about we call it `tokens_block_size` - it's only for the tokens construction, right?

**Pull request merged**